http-server/src/main/java/io/micronaut/http/server/cors/CorsFilter.java

```diff
@@ -132,12 +134,17 @@ protected boolean shouldDenyToPreventDriveByLocalhostAttack(@NonNull CorsOriginC
         if (origin == null) {
             return false;
         }
-        if (origin.startsWith(LOCALHOST)) {
+        if (isLocal(origin)) {
             return false;
         }
         String host = httpHostResolver.resolve(request);
-        return isAny(corsOriginConfiguration.getAllowedOrigins()) && host.startsWith(LOCALHOST);
+        return isAny(corsOriginConfiguration.getAllowedOrigins()) && isLocal(host);
+    }
 
+    private boolean isLocal(@NonNull String hostString) {
+        URI uri = URI.create(hostString);
+        String host = uri.getHost();
+        return "localhost".equals(host) || "127.0.0.1".equals(host);
```

Comment:
```suggestion
        return "localhost".equals(host) || host.startsWith("127");
```
this is the most accurate you get without reading the systems network settings and it matches the whole of `127.0.0.0/8`

Reply:
What happens when someone builds a website `127.0.0.malicious.com`?

Reply:
my bad not thinking about this, it would need to be coupled with something like `host.matches("\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}")`

Reply:
Yeah, I didn't want to cause a regular expression match to be executed for every request 🤔

Reply:
Using manual splitting and parsing makes it about ~30-50% *slower* from my immature and very vague performance tests.

On my local machine the test takes ~2ms with regex ~4ms with custom parsing, Still adding 2ms to a request seem less than ideal (ofc this 2ms depend on the machine/CPU, JIT and runtime warmup) i dont have a better way in mind atm

The only other thing i could think about is using the `NetworkInterface` API to check and adress for a loopback bound

The test i used
```java
var start = System.currentTimeMillis();
var rest = Arrays.stream("127.0.0.1".split("\\."))
    .allMatch(s -> {
        try{
            Integer.parseInt(s);
           return true;
        }catch(Exception e){return false;}
 });
 System.out.println(System.currentTimeMillis()-start);
```

Reply:
I'll have a think...  Have you ever seen `127.0.0.3` being used in the wild? @sdelamo any thoughts?

Reply:
Nobody uses `127.0.0.3` but it allways redirects to localhost. This means that if a webpage requests `127.200.200.200` this bypasses the check. It isn't used for serving but it is an attack surface `0.0.0.0` also includes `127.0.0.1 -> 127.254.254.254` which are 2^24 vulnerable addresses. It might prevents the simplest form of attack but any malicious actor who has any knowledge about networks can bypass it by adding 1 to the address.

Maybe i understand something wrong, do you see where i am coming from?

Reply:
Ahhh, I see now

> Because 90% use 0.0.0.0 as their api host which includes all of 127.0.0.0/8

Micronaut specifically binds to `localhost`, not 0.0.0.0 (unless the user changes the setting)

https://github.com/micronaut-projects/micronaut-core/blob/1d5016b2917e8c3467677c14e543417553be65bc/http-server-netty/src/main/java/io/micronaut/http/server/netty/NettyHttpServer.java#L399

So I don't believe 127.0.0.2 would resolve?

Reply:
Thats interesting but on my system `127.0.0.2` resolves on the default project using Arch Linux. Maybe on other systems it works different?
```bash
curl --location --request GET 'https://launch.micronaut.io/create/default/com.example.demo?lang=JAVA&build=GRADLE&test=JUNIT&javaVersion=JDK_17' --output demo.zip
unzip demo.zip
cd demo
./gradlew run
(different terminal) curl 127.0.0.3:8080
(different terminal) curl 127.200.200.203:8080
```
On my machine both curl commands pass with 404 not found json template which means the server is reached

Reply:
Can you add a reproducer as described in https://github.com/micronaut-projects/micronaut-core/pull/8642#issuecomment-1400245630

Reply:
I think osx only supports 127.0.0.1, which is why I'm not seeing it

Reply:
See : https://github.com/sdelamo/simple-request-attacks/pull/1

Reply:
I think a simple ubuntu docker container should solve your OSX reproduction problem :)